### PR TITLE
PERF: some more perf/clean in saslib.pyx

### PR DESCRIPTION
**Pull request by jreback**
more perf

cc @kshedden

### CODE REVIEW

**Comment by jreback** [comment]
@kshedden if you can run this on your example would be great (it may only be a small speedup)

**Comment by kshedden** [comment]
I'm getting ~2.99 seconds per 100K lines on the same file I have been using in #12656 (I think I was getting 3.1-3.2 seconds with the earlier version).

**Comment by kshedden** [comment]
My test file is not compressed and many of your changes are in the decompression, so it would not have helped.

**Comment by jreback** [comment]
good point

do we have a compress file to test on ? 

are those sas7cdat?

File: pandas/io/sas/saslib.pyx [hunk 1]
```diff
@@ -16,13 +19,13 @@ cdef np.ndarray[uint8_t, ndim=1] rle_decompress(int result_length, np.ndarray[ui
 
     while ipos < length:
         control_byte = inbuff[ipos] & 0xF0
-        end_of_first_byte = int(inbuff[ipos] & 0x0F)
+        end_of_first_byte = <uint8_t>(inbuff[ipos] & 0x0F)
```
Comment (kshedden) [inline pandas/io/sas/saslib.pyx]: I think we have to cast to regular int here (2 byte or wider), below there are places where end_of_first_byte is multiplied and will overflow.

Comment (jreback) [inline pandas/io/sas/saslib.pyx]: sure / do u have some test files for this ? if not ok too

**Comment by kshedden** [comment]
`end_of_first_byte` needs to be an int I think (at least 2 bytes wide).  Even though its value is copied out of a uint8 array it needs to be promoted to uint16 or it overflows later on.  We have to be very careful in the decompressors about types.  It's been a few months since I did this but there is a lot of subtle stuff in there and I spent a lot of trial and error getting it to work.  The masking and shifting gets messed up if things are not exactly the right int type (too wide or too narrow both cause problems).  I don't think this is covered well by our tests.  These are very outdated and poorly documented compression algorithms.  Most of the compression codes are rarely used except in larger files.  To keep the test suite fast I created only small test files which don't hit all the compression codes.

**Comment by kshedden** [comment]
Many of the files in our test suite are compressed, but they are small and don't necessarily use all the compression codes.  Compressed SAS files have the same name extension as decompressed ones (sas7bdat).  You might be thinking of sas7bcat which has categorical data labels (sas7bdat only supports floats and strings).

**Comment by jreback** [comment]
@kshedden ok I fixed up the byte construction. ideally we would have a test for this :>

will merge on your ok.

**Comment by jreback** [comment]
@kshedden ok with this?

**Comment by jreback** [comment]
@kshedden I merged in. lmk if any more issues (or if you have better test cases)

**Pull request closed without merging**